// serverless-python-requirements/index.js
'use strict';

const BbPromise = require('bluebird');
const fse = BbPromise.promisifyAll(require('fs-extra'));
const path = require('path');
const _ = require('lodash');
const pipCompile = require('./pipCompile');

class ServerlessPythonRequirements {
  constructor(serverless, options) {
    this.serverless = serverless;
    this.options = options;
    this.hooks = {
      'before:package:createDeploymentArtifacts': this.process.bind(this),
      'before:deploy:function:packageFunction': this.process.bind(this),
      'before:invoke:local:invoke': this.process.bind(this),
      'before:package:initialize': this.validate.bind(this),
    };
  }

  validate() {
    const runtime = this.serverless.service.provider.runtime;
    if (runtime && !runtime.startsWith('python')) {
      this.serverless.cli.log(
        'Skipping serverless-python-requirements plugin as runtime is not python.'
      );
      return BbPromise.resolve();
    }
    return BbPromise.resolve();
  }

  process() {
    return BbPromise.bind(this)
      .then(this.validate)
      .then(this.pipCompile);
  }

  pipCompile() {
    this.serverless.cli.log('Packaging Python requirements...');
    return pipCompile({ servicePath: this.serverless.config.servicePath });
  }
}

module.exports = ServerlessPythonRequirements;
